<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

```{r echo=FALSE, message=FALSE, warning=FALSE}
# What follows is code adapted from work by https://github.com/chrisroat
require(bigrquery)
require(ggplot2)
require(gridExtra)
require(reshape)
require(scales)  # for scientific_format

# Create colors that are color-blind friendly.
cbPalette = c("#000000", "#E69F00", "#56B4E9", "#009E73",
              "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
pallette_fill = scale_fill_manual(values=cbPalette)
pallette_colour = scale_colour_manual(values=cbPalette)

# Run a query against BigQuery 1k genomes table
project <- "google.com:biggene" # put your projectID here
RunQuery <- function(querySql) {
  cat(querySql)
  return(query_exec(querySql, project))
}

# Examines the quality score distibution of each variant type.
MakeQualityPlot = function() {
  data = RunQuery(paste("SELECT",
                        "  vt AS variant_type,",
                        "  quality,",
                        "  COUNT(1) AS cnt",
                        "FROM",
                        "  [genomics-public-data:1000_genomes.variants]",
                        "OMIT RECORD IF",
                        "  reference_name IN (\"X\", \"Y\", \"MT\")",
                        "GROUP BY",
                        "  variant_type,",
                        "  quality", sep="\n"))
  plot = ggplot(data, aes(x=quality, weight=cnt)) +
    geom_histogram(binwidth=25) +
      facet_grid(variant_type~., scale="free_y") +
        xlim(0, 1000) +
          xlab("Call Quality Score") +
            ylab("Count")
  return(list(plot))
}

# Examines the per-chromosome frequency of each variant type.
MakeVariantSummaryPlot = function() {
  data = RunQuery(paste("SELECT",
                        "  INTEGER(reference_name) AS chromosome,",
                        "  vt AS variant_type,",
                        "  COUNT(1) AS cnt",
                        " FROM",
                        "   [genomics-public-data:1000_genomes.variants]",
                        " OMIT RECORD IF",
                        "   reference_name IN (\"X\", \"Y\", \"MT\")",
                        " GROUP BY",
                        "   chromosome,",
                        "   variant_type", sep="\n"))
  data$variant_type <- factor(data$variant_type, levels =
                              c("SV", "INDEL", "SNP"))
  plot = ggplot(data) +
    aes(x=chromosome, y=cnt, fill=variant_type,
        order= -as.numeric(variant_type)) +
      geom_histogram(stat="identity", binwidth=1, position="fill") +
        coord_cartesian(ylim=c(0.95, 1)) +
          xlab("Chromosome Number") +
            ylab("Variant Fraction / Chromosome") +
              pallette_fill
  return(list(plot))
}

# Examine the min/max position of variants on each chromosome.
MakePositionsPlot = function() {
  data_raw = RunQuery(paste("SELECT",
                            "   INTEGER(reference_name) AS chromosome,",
                            "   MIN(start) AS min,",
                            "   MAX(start) AS max",
                            " FROM",
                            "   [genomics-public-data:1000_genomes.variants]",
                            " OMIT RECORD IF",
                            "   reference_name IN (\"X\", \"Y\", \"MT\")",
                            " GROUP BY",
                            "   chromosome", sep="\n"))
  data = melt(data_raw, id=c("chromosome"))
  plot = ggplot(data) +
    aes(x=chromosome, y=value, fill=variable) +
      geom_bar(stat="identity", position="dodge") +
        scale_y_log10() +
          xlab("Chromosome") +
            ylab("Position (log scale)")
  return(list(plot))
}

# For SNPs, plots the frequency of each pair of alleles (i.e. "A|A") given the
# reference base.
MakeSnpsPlot = function() {
  # The xxxx_allele fields are 0 if the reference base was called, and 1 if
  # the alternate base was called.  There is never more than 1 alternate
  # base for SNPs.  [A -1 is present if no call was made.]
  data = RunQuery(paste("SELECT",
                        "  reference_bases AS reference,",
                        "  CONCAT(",
                        "    IF(first_allele=0,",
                        "       reference_bases,",
                        "       alternate_bases),",
                        "    \"|\",",
                        "    IF(second_allele=0,",
                        "       reference_bases,",
                        "       alternate_bases)",
                        "    ) AS alleles,",
                        "  COUNT(1) AS cnt,",
                        "FROM",
                        "     FLATTEN((SELECT ",
                        "         reference_name,",
                        "         reference_bases,",
                        "         alternate_bases,",
                        "         vt,",
                        "         NTH(1, call.genotype) WITHIN call AS first_allele,",
                        "         NTH(2, call.genotype) WITHIN call AS second_allele",
                        "       FROM [genomics-public-data:1000_genomes.variants])",
                        "     , call)",
                        "OMIT RECORD IF",
                        "     reference_name IN (\"X\", \"Y\", \"MT\")",
                        "  OR first_allele < 0",
                        "  OR second_allele < 0",
                        "  OR vt != \"SNP\"",
                        "GROUP BY",
                        "  reference,",
                        "  alleles", sep="\n"))
  data$fill = with(data, ifelse(cnt > 1e9, 1, 0))
  hist = ggplot(data) +
    aes(x=alleles, y=cnt, fill=factor(fill)) +
      scale_fill_manual(values=c("1"="red", "0"="black"), guide=FALSE) +
        geom_histogram(stat="identity") +
          facet_grid(reference~.) +
            scale_y_log10() +
              coord_cartesian(ylim=c(1e5, 2e10)) +
                xlab("Alleles") +
                  ylab("Count")

  heatmap = ggplot(data) +
    aes(x=alleles, y=reference) +
      geom_tile(aes(fill=cnt)) +
        scale_fill_gradient2(trans="log", breaks=c(3e7, 1e8, 3e8, 1e9, 3e9))

  return(list(hist, heatmap))
}

# Plots the length distribution for Inserts/Deletes.
MakeInDelsPlot = function() {
  data = RunQuery(paste("SELECT",
                        "  CASE",
                        "    WHEN LENGTH(alternate_bases) - ",
                        "         LENGTH(reference_bases) > 50 ",
                        "      THEN 51",
                        "    WHEN LENGTH(alternate_bases) - ",
                        "         LENGTH(reference_bases) < -50 ",
                        "      THEN -51",
                        "    ELSE",
                        "      LENGTH(alternate_bases) - ",
                        "      LENGTH(reference_bases) ",
                        "END AS length,",
                        "  COUNT(1) AS cnt",
                        "FROM",
                        "     FLATTEN((SELECT ",
                        "         reference_name,",
                        "         reference_bases,",
                        "         alternate_bases,",
                        "         vt,",
                        "         NTH(1, call.genotype) WITHIN call AS first_allele,",
                        "         NTH(2, call.genotype) WITHIN call AS second_allele",
                        "       FROM [genomics-public-data:1000_genomes.variants])",
                        "     , call)",
                        "WHERE",
                        "      first_allele =",
                        "      POSITION(alternate_bases)",
                        "  AND LENGTH(alternate_bases) -",
                        "      LENGTH(reference_bases) != 0",
                        "OMIT RECORD IF",
                        "      reference_name IN (\"X\", \"Y\", \"MT\")",
                        "  AND vt != \"INDEL\"",
                        "GROUP BY",
                        "  length", sep="\n"))
  plot = ggplot(data) +
    aes(x=length, y=cnt) +
      geom_point() +
        scale_y_log10() +
          xlab("INDEL length") +
            ylab("Count (log scale)")
  return(list(plot))
}

# Makes a scatterplot, with one point per genome:
#  X: # variants with 1 allele differing from the reference
#  Y: # variants with 2 alleles differing from the reference
MakeGenomeSnpCountsPlot = function() {
  data = RunQuery(paste("SELECT",
                        "  variant_info.genome AS genome,",
                        "  CONCAT(SUBSTR(sample_info.population_description,",
                        "                0, 20), \"...\") AS population,",
                        "  sample_info.super_population_description",
                        "                AS super_population,",
                        "  SUM(variant_info.single) AS cnt1,",
                        "  SUM(variant_info.double) AS cnt2",
                        "FROM (",
                        "  FLATTEN((",
                        "  SELECT ",
                        "    call.call_set_name AS genome,",
                        "    SOME(call.genotype > 0) AND NOT EVERY(call.genotype > 0) WITHIN call AS single,",
                        "    EVERY(call.genotype > 0) WITHIN call AS double,",
                        "   FROM [genomics-public-data:1000_genomes.variants]",
                        "   OMIT RECORD IF",
                        "     reference_name IN (\"X\", \"Y\", \"MT\"))",
                        "  , call)",
                        "  ) AS variant_info",
                        "  JOIN",
                        "    [genomics-public-data:1000_genomes.sample_info] AS sample_info",
                        "  ON",
                        "    variant_info.genome = sample_info.sample",
                        "GROUP BY",
                        "  genome,",
                        "  population,",
                        "  super_population", sep="\n"))
  pop_plot = ggplot(data) +
    aes(x=cnt1, y=cnt2, color=population) +
      geom_point(shape=1, size=1) +
        scale_x_continuous(label=scientific_format()) +
          scale_y_continuous(label=scientific_format()) +
            xlab("SNPs with 1 Diff. Allele") +
              ylab("SNPs with 2 Diff. Allele")
  super_pop_plot = ggplot(data) +
    aes(x=cnt1, y=cnt2, color=super_population) +
      geom_point(shape=1, size=1) +
        scale_x_continuous(label=scientific_format()) +
          scale_y_continuous(label=scientific_format()) +
            xlab("SNPs with 1 Diff. Allele") +
              ylab("SNPs with 2 Diff. Allele")
  return(list(pop=pop_plot, super_pop=super_pop_plot))
}

# Likelhoods for unphased variants.
# TODO(roat): Make sure timezones are handled correctly and include phased variants.
MakeLikelihoodPlot = function() {
  data = RunQuery(paste("SELECT",
                        "  variant_type,",
                        "  likelihood,",
                        "  COUNT(1) AS cnt",
                        "FROM (",
                        "  SELECT",
                        "    variant_type,",
                        "    ROUND(100 * IF(gl > -0.5, gl,",
                        "          -0.5)) AS likelihood,",
                        "  FROM",
                        "     FLATTEN((SELECT ",
                        "         reference_name,",
                        "         vt AS variant_type,",
                        "         call.call_set_name AS genome,",
                        "         call.phaseset AS phaseset,",
                        "         call.genotype_likelihood AS gl,",
                        "         NTH(1, call.genotype) WITHIN call AS first_allele,",
                        "         NTH(2, call.genotype) WITHIN call AS second_allele",
                        "       FROM [genomics-public-data:1000_genomes.variants])",
                        "     , call)",
                        "  WHERE",
                        "        (first_allele <= second_allele",
                        "          AND POSITION(gl) = 1 +",
                        "          (second_allele *",
                        "            (second_allele + 1) / 2) +",
                        "          first_allele)",
                        "        OR (second_allele < first_allele",
                        "          AND POSITION(gl) = 1 +",
                        "          (first_allele *",
                        "            (first_allele + 1) / 2) +",
                        "          second_allele)",
                        "  OMIT RECORD IF ",
                        "       reference_name in (\"X\", \"Y\", \"MT\")",
                        "    OR phaseset IS NULL",
                        ")",
                        "GROUP BY",
                        "  variant_type,",
                        "  likelihood", sep="\n"))
  plot = ggplot(data) +
    aes(x=likelihood, weight=cnt) +
      geom_histogram(binwidth=2) +
        facet_grid(variant_type~., scale="free_y") +
          xlab("100 * Likelihood") +
            ylab("Count")
  return(list(plot))
}

# Outputs the a png.
CreateGriddedPlot = function(plots) {
  grob = do.call(arrangeGrob, plots)
  grid.draw(grob)
}
```

Basic Plots of 1k Genome Data
=============================

Min/Max Chromosomal Positions of Variants
-----------------------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakePositionsPlot()
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(plots)
```
 * A basic sanity check

 * Chromosomes 13,14,15,21,22 have abnormally high min positions.
 * Chromosome 7 has low min position.
 * Hard to sequence regions?

Frequency of Variant Types Per Chromosome
-----------------------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakeVariantSummaryPlot()
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(plots)
```
 * Mostly SNPs.
 * Very few structural variants.
 * Note suppressed zero.

Types of SNP pairs (looks at both alleles)
------------------------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakeSnpsPlot()
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(plots)
```
 * Total count is #genomes * #SNPs.
 * Same data - two views

 * Red bars denote no mutation - both alleles equal reference.

 * Why are A -> C,C and A-> G,G likely, but not A -> C,G or A -> G,C?

 * Note that A & T rows are reverses of each other, as are C & G rows.

Length of Insertion/Deletion
---------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakeInDelsPlot()
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(plots)
```
 * +/-51 are over/under flow bins
 * -> Large tail of deletions

 * Positive length = Insertion
 * Negative length = Deletion

 * Drops off quickly.

Quality score of calls (at least, of INDELs)
--------------------------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakeQualityPlot()
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(plots)
```
From the 1k genome docs:
> phred-scaled quality score for the assertion made in ALT. i.e. -10log_10 prob(call in ALT is wrong). If ALT is ”.” (no variant) then this is -10log_10 p(variant), and if ALT is not ”.” this is -10log_10 p(no variant). High QUAL scores indicate high confidence calls.

From Broad Institute:
> The Phred scaled probability that a REF/ALT polymorphism exists at this site given sequencing data. Because the Phred scale is -10 * log(1-p), a value of 10 indicates a 1 in 10 chance of error, while a 100 indicates a 1 in 10^10 chance. These values can grow very large when a large amount of NGS data is used for variant calling.

Likelihood Scores for each Allele
---------------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakeLikelihoodPlot()
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(plots)
```
This is the likelihood for the most likely set of alleles for each variant.

SNP distribution in Genomes
---------------------------
```{r echo=FALSE, message=FALSE, warning=FALSE, error=FALSE, cache=FALSE, comment=NA}
plots <- MakeGenomeSnpCountsPlot()
```
First by Population:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(list(plots$pop))
```
Each point is a genome:
 * X coord denotes the #SNPs w/ 1 mutation
 * Y coord denotes the #SNPs w/ 2 mutations

Cluster correlate very well with ethnicity.

Then by super population:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align="center", fig.width=12, fig.height=8}
CreateGriddedPlot(list(plots$super_pop))
```
