<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

Annotation JOINs
========================================================

In this data story we explore JOINing variant data in BigQuery with several annotation databases.  **NOTE: the JOIN criteria may not be correct.**  Please only use this resource for ideas, and file github issues for corrections.


```{r}
require(bigrquery)
require(ggplot2)
require(xtable)
require(testthat)
project <- "google.com:biggene" # put your projectID here
DisplayAndDispatchQuery <- function(queryUri) {
  querySql <- readChar(queryUri, nchars=1e6)
  cat(querySql)
  query_exec(querySql, project)
}
```

Let us start by getting an idea of the number of rows in each of our tables.
```{r cache=FALSE}
tables <- c("1000genomes.phase1_variants",
            "1000genomes.pedigree",
            "1000genomes.sample_info",
            "annotations.clinvar",
            "annotations.clinvar_disease_names",
            "annotations.known_genes",
            "annotations.known_genes_aliases")
lapply(tables, function(table) {
  querySql <- paste0("SELECT count(*) AS cnt FROM [google.com:biggene:",
                table, "]")
  result <- query_exec(querySql, project)
  paste(table, result, sep=": ")
})
```
We can see that all the annotation databases are dwarfed in size by the data in the variants table.

Let us also examine the types of variants within ClinVar.
```{r cache=FALSE}
result <- query_exec("SELECT type, count(1) cnt FROM [google.com:biggene:annotations.clinvar] group by type",
                     project)
```
Number of rows returned by this query: `r nrow(result)`.

Display the rows of our result
```{r comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
We can see that the vast majority of variants within ClinVar are SNPs.

## JOINs that check whether a position overlaps an interval

### JOINing Sample SNP Variants with ClinVar

```{r cache=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/individual-clinically-concerning-variants.sql")
```
Number of rows returned by this query: `r nrow(result)`.

Display the first few rows of our result
```{r comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
We can see that this indivudual has 53 clinically concerning variants.

### JOINing Sample SNP Variants with ClinVar, Grouped by Family
```{r cache=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/familial-shared-clinically-concerning-variants.sql")
```
Number of rows returned by this query: `r nrow(result)`.

Display the first few rows of our result
```{r comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
We can see that some variants are shared by as many as four family members.

### JOINing Sample INDEL Variants with ClinVar

First, let us see what the INDEL data looks like in ClinVar:
```{r cache=FALSE}
querySql <- 'SELECT * FROM [google.com:biggene:annotations.clinvar] where type="indel"'
result <- query_exec(querySql, project)
```
Number of rows returned by this query: `r nrow(result)`.

Display the first few rows of our result
```{r comment=NA, results="asis"}
print(xtable(head(result, 20)), type="html", include.rownames=F)
```
We can see that its not obvious as to how to parse the the ClinVar INDEL details in order to know which INDELs within 1,000 Genomes might match.

## JOINs that check whether an interval overlaps another interval

### JOINing Chromosome 17 Variants with Gene Names

Next we will JOIN our variants with gene names.  Note that the JOIN criteria is simple - just matching on the chromosome, but the WHERE clause ensures the intervals overlap.
```{r cache=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/gene-variant-counts.sql")
```
Number of rows returned by this query: `r nrow(result)`.

Display the first few rows of our result
```{r comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
And drilling down to just the genes with name matching BRCA1
```{r}
brca1_all <- subset(result, grepl("BRCA1", gene_aliases))
dim(brca1_all)
```

```{r comment=NA, results="asis"}
print(xtable(brca1_all), type="html", include.rownames=F)
```
We see how many variants we have within these genes for the full dataset.

### JOINing Chromosome 17 Variants for a Particular Sample with Gene Names

Now let us look at these sample variants for a particular sample
```{r cache=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/sample-gene-variant-counts.sql")
```
Number of rows returned by this query: `r nrow(result)`.

Display the first few rows of our result
```{r comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
And drilling down to just the genes with name matching BRCA1
```{r}
brca1_one <- subset(result, grepl("BRCA1", gene_aliases))
dim(brca1_one)
```

```{r comment=NA, results="asis"}
print(xtable(brca1_one), type="html", include.rownames=F)
```

Let us compare these to the dataset level gene counts
```{r all vs one}
brca1_all$name == brca1_one$name
expect_that(brca1_all$name, equals(brca1_one$name))
brca1_all$cnt - brca1_one$cnt
mean(brca1_all$cnt - brca1_one$cnt)
qplot(brca1_all$cnt, brca1_one$cnt,
      xlim=c(0, max(brca1_all$cnt)),
      ylim=c(0, max(brca1_all$cnt)),
      xlab="count of variants per gene for the full dataset",
      ylab="count of variants per gene for one sample",
      )
```
And we see that our sample has variants within the same set of genes, but many fewer per gene.

### JOINing All Variants with Gene Names
Let us go bigger now and run this on the entire 1,000 Genomes dataset.
```{r cache=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../../sql/specific-gene-variant-counts.sql")
```
Number of rows returned by this query: `r nrow(result)`.

Display the rows of our result
```{r comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```
And we see the count of variants in this entire dataset found within the genes corresponding to BRCA1 and APOE.
